Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -132,63 +132,75 @@ static void deleteEmptyOriginDirectory(const String& directory)
 {
-    return adoptRef(*new NetworkStorageManager(sessionID, identifier, connection, path, customLocalStoragePath, customIDBStoragePath, customCacheStoragePath, defaultOriginQuota, defaultThirdPartyOriginQuota, originQuotaRatio, totalQuotaRatio, volumeCapacityOverride, level));
+    auto manager = adoptRef(*new NetworkStorageManager(sessionID, identifier, connection, path));
+    manager->initialize(path, customLocalStoragePath, customIDBStoragePath, customCacheStoragePath, defaultOriginQuota, defaultThirdPartyOriginQuota, originQuotaRatio, totalQuotaRatio, volumeCapacityOverride, level);
+    return manager;
 }
 
-NetworkStorageManager::NetworkStorageManager(PAL::SessionID sessionID, Markable<UUID> identifier, IPC::Connection::UniqueID connection, const String& path, const String& customLocalStoragePath, const String& customIDBStoragePath, const String& customCacheStoragePath, uint64_t defaultOriginQuota, uint64_t defaultThirdPartyOriginQuota, std::optional<double> originQuotaRatio, std::optional<double> totalQuotaRatio, std::optional<uint64_t> volumeCapacityOverride, UnifiedOriginStorageLevel level)
-    : m_sessionID(sessionID)
-    , m_queueName(makeString("com.apple.WebKit.Storage.", sessionID.toUInt64(), ".", static_cast<uint64_t>(identifier->data() >> 64), static_cast<uint64_t>(identifier->data())))
-    , m_queue(SuspendableWorkQueue::create(m_queueName.utf8().data(), SuspendableWorkQueue::QOS::Default, SuspendableWorkQueue::ShouldLog::Yes))
-    , m_defaultOriginQuota(defaultOriginQuota)
-    , m_defaultThirdPartyOriginQuota(defaultThirdPartyOriginQuota)
-    , m_originQuotaRatio(originQuotaRatio)
-    , m_totalQuotaRatio(totalQuotaRatio)
-    , m_volumeCapacityOverride(volumeCapacityOverride)
-    , m_parentConnection(connection)
+void NetworkStorageManager::initialize(const String& path, const String& customLocalStoragePath, const String& customIDBStoragePath, const String& customCacheStoragePath, uint64_t defaultOriginQuota, uint64_t defaultThirdPartyOriginQuota, std::optional<double> originQuotaRatio, std::optional<double> totalQuotaRatio, std::optional<uint64_t> volumeCapacityOverride, UnifiedOriginStorageLevel level)
+{
+    m_queue->dispatch([this, protectedThis = Ref { *this }, path = path.isolatedCopy(), customLocalStoragePath = customLocalStoragePath.isolatedCopy(), customIDBStoragePath = customIDBStoragePath.isolatedCopy(), customCacheStoragePath = customCacheStoragePath.isolatedCopy(), defaultOriginQuota, defaultThirdPartyOriginQuota, originQuotaRatio, totalQuotaRatio, volumeCapacityOverride, level]() mutable {
```

Comment:
Could we have captured a ThreadSafeWeakPtr instead and avoided the refactoring?

Reply:
Yes, it is safe to create WeakPtr or TheadSafeWeakPtr from inside constructors.

Also note that it is generally safe to ref on object from inside its constructor as far as I know. However, you'd need to relax adoption requirements (there is a function could can call for that) to avoid hitting the assertion in debug. The point of the assertion wasn't to stop you from ref'ing the object from inside the constructor. Its point is to make sure that people don't forget to adopt newly constructed refcounted objects.

Anyway, here, my recommendation would be to use ThreadSafeWeakPtr<> since I don't think we truly need to extend the lifetime and since it looks a little cleaner than relaxing adoption requirements.